tiledb/subarray.py

```diff
@@ -29,24 +21,33 @@ class Subarray(CtxMixin, lt.Subarray):
     def __init__(
         self,
         array: Array,
-        ctx: "Ctx" = None,
+        ctx: Ctx = None,
     ):
         """Class representing a subarray of a TileDB Array.
 
         :param array: tiledb.Array the subarray is defined on
         :param ctx: A TileDB Context
         """
         self._array = array
-        self._cpp_array = ArrayImpl(array)
-        super().__init__(ctx, self._cpp_array)
+        super().__init__(ctx, ArrayWrapper(array))
 
-    def nrange(self, key) -> np.uint64:
-        """Returns the number of ranges on a dimension.
+    def add_dim_range(self, dim_idx: int, dim_range: Range):
+        """Add a range to a dimension of the subarray.
 
-        :param key: dimension index (int) or name (str)
-        :rtype: np.uint64
+        :param dim_idx: dimension index (int) of the dimension to add the range to
+        :param range: tuple containing the inclusive range to query on
+        :raises: :py:exc:`tiledb.TileDBError`
         """
-        return self._range_num(key)
+        self._add_dim_range(dim_idx, dim_range)
```

Comment:
Except for `__init__` and `add_ranges`, every other method in this class just delegates to the respective PyBind11 method. Is there a convention to have all public methods implemented in Python (or equivalently, all C++ methods be private), e.g for easier documentation?

Reply:
AFAIK there's nothing that requires or encourages having "private" parent class/methods in PyBind11 and public child ones in pure Python. I'm not too keen on extra levels of indirection without a concrete benefit but I'll defer to @nguyenv and @ihnorton.

Reply:
It looks like you can actually [add docstrings to the pybind class](https://pybind11.readthedocs.io/en/stable/advanced/misc.html#generating-documentation-using-sphinx). I'll remove the wrappers functions and make the pybind methods public in a future PR.